Don't bundle NativeAOT runtime assets in the ILCompiler package

jkoritzinsky opened the pull request:
With https://github.com/dotnet/sdk/pull/46611, the .NET SDK will now use the "NativeAOT runtime pack" as the "runtime package" when publishing for NativeAOT. It's not as clean as https://github.com/dotnet/sdk/pull/37872, but it has the same effect on our packaging.

Now that the .NET SDK will always look in the "NativeAOT runtime pack" for .NET 10 and newer, we can do two things:

1. Don't bundle the runtime assets in the rid-specific ILCompiler package.
2. Remove support in the ILCompiler MSBuild targets for the old package layout.

We leave the different properties as-is (ie the split between `aotsdk` and `framework`) for usage in our tests.
 
Contributes to https://github.com/dotnet/runtime/issues/87060

jkoritzinsky committed:
Don't bundle NativeAOT runtime assets in the ILCompiler package

jkoritzinsky committed:
Remove support for the old package layout from the targets.

jkoritzinsky commented:
I need to update those docs to match the changes, but using it should still be possible (if slightly more annoying).

jkoritzinsky committed:
Update docs to enable downloading the runtime pack manually during restore (as otherwise the right version number isn't available until after restore).

jkotas commented on docs/workflow/building/coreclr/nativeaot.md:
```diff
@@ -34,6 +34,14 @@ Run `build[.cmd|.sh] -c Release` from the repo root to build the NativeAOT toolc
 
 * Add the package directory to your `nuget.config` file. For example, add `<add key="local" value="C:\runtime\artifacts\packages\Release\Shipping" />`
 * Run `dotnet add package Microsoft.DotNet.ILCompiler -v 10.0.0-dev` to add the local package reference to your project.
```
```suggestion
```

jkotas commented on docs/workflow/building/coreclr/nativeaot.md:
````diff
@@ -34,6 +34,14 @@ Run `build[.cmd|.sh] -c Release` from the repo root to build the NativeAOT toolc
+
+```
+<ItemGroup>
+   <PackageDownload Include="Microsoft.NETCore.App.Runtime.NativeAOT.win-x64" Version="[10.0.0-dev]" />
````
```suggestion
   <PackageReference Include="Microsoft.DotNet.ILCompiler" Version="10.0.0-dev" />
   <PackageDownload Include="Microsoft.NETCore.App.Runtime.NativeAOT.win-x64" Version="[10.0.0-dev]" />
```

am11 commented:
> Don't bundle the runtime assets in the rid-specific ILCompiler package.

Are they bundled today? I don't see it:

```
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/Icon.png
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/runtime.json
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/tools
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/tools/netstandard
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/tools/netstandard/ILCompiler.Build.Tasks.dll
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/tools/netstandard/ILCompiler.Build.Tasks.deps.json
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/.nupkg.metadata
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/microsoft.dotnet.ilcompiler.nuspec
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/microsoft.dotnet.ilcompiler.10.0.0-preview.6.25304.106.nupkg
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/THIRD-PARTY-NOTICES.TXT
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build/WindowsAPIs.txt
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build/Microsoft.NETCore.Native.Unix.targets
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build/NativeAOT.natstepfilter
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build/Microsoft.NETCore.Native.Publish.targets
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build/BuildFrameworkNativeObjects.proj
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build/Microsoft.DotNet.ILCompiler.SingleEntry.targets
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build/findvcvarsall.bat
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build/Microsoft.NETCore.Native.targets
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build/NativeAOT.natvis
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build/Microsoft.NETCore.Native.Windows.targets
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/build/Microsoft.DotNet.ILCompiler.props
microsoft.dotnet.ilcompiler/10.0.0-preview.6.25304.106/microsoft.dotnet.ilcompiler.10.0.0-preview.6.25304.106.nupkg.sha512
```

`runtime.${RID}.microsoft.dotnet.ilcompile` is already a separate package.

filipnavara commented:
> `runtime.${RID}.microsoft.dotnet.ilcompile` is already a separate package.

The problem is cross-compiling where the ILCompiler is on the host platform (eg. osx-arm64) and you are targeting a different platform (eg. ios-arm64). Then you'd get `runtime.osx-arm64.microsoft.dotnet.ilcompiler` with all the SDK assets that are unused because they are taken from the target pack for ios-arm64.

jkotas commented on docs/workflow/building/coreclr/nativeaot.md:
````diff
@@ -34,6 +34,14 @@ Run `build[.cmd|.sh] -c Release` from the repo root to build the NativeAOT toolc
+<ItemGroup>
+   <PackageDownload Include="Microsoft.NETCore.App.Runtime.NativeAOT.win-x64" Version="[10.0.0-dev]" />
+</ItemGroup>
+```
````
Similar instructions are also at https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/docs/compiling.md?plain=1#L20-L32 . They need to be updated as well.

jkotas commented:
This will impact number of places outside this repo that reference specific NAOT version via package reference. For example:

https://github.com/dotnet/dotnet/blob/main/src/aspnetcore/eng/tools/GenerateFiles/Directory.Build.targets.in#L60-L62
https://github.com/dotnet/BenchmarkDotNet/blob/master/src/BenchmarkDotNet/Toolchains/NativeAot/Generator.cs#L174-L175
https://github.com/aspnet/Benchmarks/blob/main/build/nativeaot-scenarios.yml#L40-L56

Is there really no way to get SDK to add the matching `PackageDownload` implicitly so that we do not have to chase down all these places or debug issues caused by mismatched versions between the two lines?

jkoritzinsky commented:
The main problem is the ILCompiler targets exist in dotnet/runtime.

I think I can figure out a way to get this working in dotnet/sdk (introducing a new property that will likely work cleaner anyway).

I'll put out a PR to dotnet/sdk after I get back.

jkoritzinsky committed:
Remove unneeded changes

jkoritzinsky committed:
Merge branch 'main' into ilcompiler-host-only

akoeplinger committed:
Merge branch 'main' into ilcompiler-host-only

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into ilcompiler-host-only

jkoritzinsky committed:
Don't pull CoreLib from the native path. Let it get pulled from the AOT SDK path.

These are the same in the standard case, but when doing local testing, the AOT SDK path has the right CoreLib, but the Framework Native path has the wrong one.

jkoritzinsky commented:
/ba-g weird helix issue.
